File: lib/scorecard.ts
/**
 * Weekly training scorecard + check-in decision rules (adaptive plan Phase 3).
 * Pure, client-safe math like lib/running.ts — zero AI, zero I/O, fully
 * unit-testable (see scripts/scorecard-check.ts).
 */

import { normalizeLoggedExercises } from "./workout-sets";

export type WeekScorecard = {
  /** Completed non-meal_note items / planned non-meal_note items, 0-100. */
  adherence_pct: number;
  planned_items: number;
  completed_items: number;
  planned_km: number;
  actual_km: number;
  /** Session-log notes/messages flagged "red" by the feedback layer. */
  red_flags: string[];
  /** Session-log notes/messages flagged "caution". */
  caution_flags: string[];
};

export type CheckinDecision = "advance" | "repeat" | "deload";

export type ScorecardItem = {
  item_type: string;
  is_completed: boolean;
  details: Record<string, unknown> | null;
};

export type ScorecardSessionLog = {
  actual: Record<string, unknown> | null;
  ai_feedback: Record<string, unknown> | null;
  note: string | null;
};

function round1(value: number): number {
  return Math.round(value * 10) / 10;
}

function numberOrNull(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) && value > 0
    ? value
    : null;
}

/**
 * Deterministic weekly scorecard. `sessionLogs[i]` is the log for `items[i]`
 * (null/undefined when the item has no log) — pass them index-aligned so the
 * actual-km fallback can pair a completed run with its missing log.
 */
export function computeWeekScorecard(
  items: ScorecardItem[],
  sessionLogs: (ScorecardSessionLog | null | undefined)[],
): WeekScorecard {
  const trainable = items.filter((item) => item.item_type !== "meal_note");
  const planned = trainable.length;
  const completed = trainable.filter((item) => item.is_completed).length;
  // 0 planned items means nothing could be missed — treat as full adherence.
  const adherence =
    planned === 0 ? 100 : round1((completed / planned) * 100);

  let plannedKm = 0;
  let actualKm = 0;
  items.forEach((item, index) => {
    const log = sessionLogs[index] ?? null;
    if (item.item_type === "run") {
      const plannedDistance = numberOrNull(item.details?.distance_km);
      if (plannedDistance !== null) plannedKm += plannedDistance;
      const loggedDistance = numberOrNull(log?.actual?.distance_km);
      if (loggedDistance !== null) {
        actualKm += loggedDistance;
      } else if (item.is_completed && plannedDistance !== null) {
        // Completed run without a logged distance counts as planned distance.
        actualKm += plannedDistance;
      }
    }
  });

  const redFlags: string[] = [];
  const cautionFlags: string[] = [];
  for (const log of sessionLogs) {
    if (!log?.ai_feedback) continue;
    const flag = String(log.ai_feedback.flag ?? "");
    if (flag !== "red" && flag !== "caution") continue;
    const text =
      (typeof log.note === "string" && log.note.trim()) ||
      (typeof log.ai_feedback.message === "string" &&
        log.ai_feedback.message.trim()) ||
      "Flagged session";
    (flag === "red" ? redFlags : cautionFlags).push(text);
  }

  return {
    adherence_pct: adherence,
    planned_items: planned,
    completed_items: completed,
    planned_km: round1(plannedKm),
    actual_km: round1(actualKm),
    red_flags: redFlags,
    caution_flags: cautionFlags,
  };
}

/**
 * Hypertrophy stall detection (adaptive plan Phase 5): an exercise logged in
 * each of the last 3 weeks with no weight OR rep increase from first to last
 * is stalled — surface at check-in to suggest a deload or variation.
 * Pass per-week strength logs oldest → newest; needs ≥3 weeks of data.
 */
export function detectStalledLifts(
  weeklyStrengthLogs: (ScorecardSessionLog | null | undefined)[][],
): string[] {
  if (weeklyStrengthLogs.length < 3) return [];
  const lastThree = weeklyStrengthLogs.slice(-3);

  // Per week: exercise name → best (weight, reps) seen that week.
  // normalizeLoggedExercises accepts both the legacy flat rows and the
  // per-set shape logged since the Hevy-style editor.
  const weekMaps = lastThree.map((logs) => {
    const map = new Map<string, { weight: number; reps: number }>();
    for (const log of logs) {
      for (const exercise of normalizeLoggedExercises(log?.actual?.exercises)) {
        const name = exercise.name.toLowerCase();
        for (const set of exercise.sets) {
          const weight = numberOrNull(set.weight_kg) ?? 0;
          const reps = numberOrNull(set.reps) ?? 0;
          const best = map.get(name);
          if (!best || weight > best.weight || (weight === best.weight && reps > best.reps)) {
            map.set(name, { weight, reps });
          }
        }
      }
    }
    return map;
  });

  const stalled: string[] = [];
  for (const [name, first] of weekMaps[0]) {
    const mid = weekMaps[1].get(name);
    const last = weekMaps[2].get(name);
    if (!mid || !last) continue;
    if (last.weight <= first.weight && last.reps <= first.reps) {
      stalled.push(name);
    }
  }
  return stalled;
}

/**
 * Rules-first check-in decision (notes Decision 2): red flag → deload;
 * two consecutive weeks under 50% adherence → deload; one week under 50%
 * → repeat; otherwise advance.
 */
export function decideWeek(
  current: WeekScorecard,
  previous: WeekScorecard | null,
): { decision: CheckinDecision; reasons: string[] } {
  if (current.red_flags.length > 0) {
    return {
      decision: "deload",
      reasons: [
        `A session was red-flagged: "${current.red_flags[0]}" — easing off to recover.`,
      ],
    };
  }
  if (current.adherence_pct < 50) {
    if (previous && previous.adherence_pct < 50) {
      return {
        decision: "deload",
        reasons: [
          `Two weeks in a row under 50% adherence (${previous.adherence_pct}% then ${current.adherence_pct}%) — a lighter deload week rebuilds momentum.`,
        ],
      };
    }
    return {
      decision: "repeat",
      reasons: [
        `Only ${current.completed_items} of ${current.planned_items} sessions done (${current.adherence_pct}%) — repeating the week instead of advancing.`,
      ],
    };
  }
  const reasons = [
    `${current.completed_items} of ${current.planned_items} sessions done (${current.adherence_pct}%) — on track to advance.`,
  ];
  if (current.caution_flags.length > 0) {
    reasons.push(
      `Heads up: ${current.caution_flags.length} session${current.caution_flags.length === 1 ? "" : "s"} flagged for caution.`,
    );
  }
  return { decision: "advance", reasons };
}

File: apps/api/internal/domain/scorecard/scorecard.go
// Package scorecard is the weekly training scorecard, the check-in decision
// rules, and hypertrophy stall detection (FORMULAS.md §7). Rules only — no AI.
package scorecard

import (
	"fmt"
	"strings"

	"github.com/hamid-karimi/coachin/apps/api/internal/domain/jsnum"
	"github.com/hamid-karimi/coachin/apps/api/internal/domain/workout"
)

// Item is the part of a plan item the scorecard reads.
type Item struct {
	ItemType    string         `json:"item_type"`
	IsCompleted bool           `json:"is_completed"`
	Details     map[string]any `json:"details"`
}

// SessionLog is the part of a session log the scorecard reads.
type SessionLog struct {
	Actual     map[string]any `json:"actual"`
	AIFeedback map[string]any `json:"ai_feedback"`
	Note       *string        `json:"note"`
}

// Week is one plan week's scorecard.
type Week struct {
	// AdherencePct is completed / planned non-meal items, 0–100, one decimal.
	AdherencePct   float64  `json:"adherence_pct"`
	PlannedItems   int      `json:"planned_items"`
	CompletedItems int      `json:"completed_items"`
	PlannedKm      float64  `json:"planned_km"`
	ActualKm       float64  `json:"actual_km"`
	RedFlags       []string `json:"red_flags"`
	CautionFlags   []string `json:"caution_flags"`
}

func round1(x float64) float64 { return jsnum.Round(x*10) / 10 }

// positive reads a JSON number > 0 (strings don't count).
func positive(v any) (float64, bool) {
	n, ok := v.(float64)
	return n, ok && n > 0
}

// ComputeWeek scores a week. logs[i] is items[i]'s session log (nil when it
// has none), so a completed run without a logged distance counts its planned
// distance.
func ComputeWeek(items []Item, logs []*SessionLog) Week {
	week := Week{AdherencePct: 100, RedFlags: []string{}, CautionFlags: []string{}}

	var plannedKm, actualKm float64
	for i, item := range items {
		if item.ItemType != "meal_note" {
			week.PlannedItems++
			if item.IsCompleted {
				week.CompletedItems++
			}
		}
		if item.ItemType != "run" {
			continue
		}
		planned, hasPlanned := positive(item.Details["distance_km"])
		if hasPlanned {
			plannedKm += planned
		}
		var log *SessionLog
		if i < len(logs) {
			log = logs[i]
		}
		if log != nil {
			if logged, ok := positive(log.Actual["distance_km"]); ok {
				actualKm += logged
				continue
			}
		}
		if item.IsCompleted && hasPlanned {
			actualKm += planned
		}
	}
	// No planned items means nothing could be missed: full adherence.
	if week.PlannedItems > 0 {
		week.AdherencePct = round1(float64(week.CompletedItems) / float64(week.PlannedItems) * 100)
	}
	week.PlannedKm, week.ActualKm = round1(plannedKm), round1(actualKm)

	flagged := map[string]*[]string{"red": &week.RedFlags, "caution": &week.CautionFlags}
	for _, log := range logs {
		if log == nil || log.AIFeedback == nil {
			continue
		}
		flag := ""
		if v := log.AIFeedback["flag"]; v != nil {
			flag = jsnum.ToString(v)
		}
		list, ok := flagged[flag]
		if !ok {
			continue
		}
		*list = append(*list, flagText(log))
	}
	return week
}

// flagText is the note, else the AI message, else a generic label.
func flagText(log *SessionLog) string {
	if log.Note != nil {
		if note := jsnum.Trim(*log.Note); note != "" {
			return note
		}
	}
	if message, ok := log.AIFeedback["message"].(string); ok {
		if trimmed := jsnum.Trim(message); trimmed != "" {
			return trimmed
		}
	}
	return "Flagged session"
}

// Decision is what a check-in does with next week.
type Decision string

// Check-in decisions.
const (
	Advance Decision = "advance"
	Repeat  Decision = "repeat"
	Deload  Decision = "deload"
)

// Result is a decision with the human-readable reasons behind it.
type Result struct {
	Decision Decision `json:"decision"`
	Reasons  []string `json:"reasons"`
}

// Decide applies the rules, first match wins: a red flag → deload; two weeks
// in a row under 50% adherence → deload; one → repeat; otherwise advance.
func Decide(current Week, previous *Week) Result {
	pct := jsnum.FormatNumber
	if len(current.RedFlags) > 0 {
		return Result{Deload, []string{
			// Plain quotes, not %q: the note is shown verbatim, never escaped.
			`A session was red-flagged: "` + current.RedFlags[0] + `" — easing off to recover.`,
		}}
	}
	if current.AdherencePct < 50 {
		if previous != nil && previous.AdherencePct < 50 {
			return Result{Deload, []string{fmt.Sprintf(
				"Two weeks in a row under 50%% adherence (%s%% then %s%%) — a lighter deload week rebuilds momentum.",
				pct(previous.AdherencePct), pct(current.AdherencePct))}}
		}
		return Result{Repeat, []string{fmt.Sprintf(
			"Only %d of %d sessions done (%s%%) — repeating the week instead of advancing.",
			current.CompletedItems, current.PlannedItems, pct(current.AdherencePct))}}
	}
	reasons := []string{fmt.Sprintf("%d of %d sessions done (%s%%) — on track to advance.",
		current.CompletedItems, current.PlannedItems, pct(current.AdherencePct))}
	if n := len(current.CautionFlags); n > 0 {
		plural := "s"
		if n == 1 {
			plural = ""
		}
		reasons = append(reasons, fmt.Sprintf("Heads up: %d session%s flagged for caution.", n, plural))
	}
	return Result{Advance, reasons}
}

type best struct{ weight, reps float64 }

// StalledLifts names exercises logged in each of the last three weeks with no
// weight or rep increase from the first of those weeks to the last. Pass
// per-week strength logs oldest → newest; fewer than three weeks → none.
func StalledLifts(weeks [][]*SessionLog) []string {
	stalled := []string{}
	if len(weeks) < 3 {
		return stalled
	}
	lastThree := weeks[len(weeks)-3:]
	bests := make([]map[string]best, 3)
	var firstWeekOrder []string // names in first-logged order

	for w, logs := range lastThree {
		bests[w] = map[string]best{}
		for _, log := range logs {
			var raw any
			if log != nil {
				raw = log.Actual["exercises"]
			}
			for _, exercise := range workout.NormalizeLoggedExercises(raw) {
				name := strings.ToLower(exercise.Name)
				for _, set := range exercise.Sets {
					weight, _ := positive(set.WeightKg)
					reps, _ := positive(float64(set.Reps))
					current, seen := bests[w][name]
					if !seen && w == 0 {
						firstWeekOrder = append(firstWeekOrder, name)
					}
					if !seen || weight > current.weight || (weight == current.weight && reps > current.reps) {
						bests[w][name] = best{weight, reps}
					}
				}
			}
		}
	}

	for _, name := range firstWeekOrder {
		first := bests[0][name]
		_, inMid := bests[1][name]
		last, inLast := bests[2][name]
		if inMid && inLast && last.weight <= first.weight && last.reps <= first.reps {
			stalled = append(stalled, name)
		}
	}
	return stalled
}
